ballerina/variable_validator.bal

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+//
+// WSO2 Inc. licenses this file to you under the Apache License,
+// Version 2.0 (the "License"); you may not use this file except
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class VariableValidator{
+    *parser:Visitor;
+
+    private parser:DocumentNode documentNode;
+    private map<json> variables;
+    private string[] visitedVariableDefinitions;
+    private ErrorDetail[] errors;
+
+    public isolated function init(parser:DocumentNode documentNode, map<json>? variableValues){
+        self.documentNode = documentNode;
+        self.visitedVariableDefinitions = [];
+        self.errors = [];
+        if variableValues == () {
+            self.variables = {};
+        } else {
+            self.variables = <map<json>> variableValues;
+        }
+    }
+
+    public isolated function validate() returns ErrorDetail[]? {
+        self.visitDocument(self.documentNode);
+        if self.errors.length() > 0 {
+            return self.errors;
+        }
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode, anydata data = ()) {
+        parser:OperationNode[] operations = documentNode.getOperations();
+        foreach parser:OperationNode operationNode in operations {
+            foreach ErrorDetail errorDetail in operationNode.getErrors() {
+                self.errors.push(errorDetail);
+            }
+            self.visitOperation(operationNode, operationNode.getVaribleDefinitions());
+        }
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode, anydata data = ()) {
+        map<parser:VariableDefinition> variableDefinitions = <map<parser:VariableDefinition>> data;
```

Comment:
If we directly access the variable definitions from the operation node, this casting can be avoided.